marlowe-runtime-cli/app/Language/Marlowe/Runtime/CLI/Command/Load.hs

```diff
@@ -0,0 +1,322 @@
+import Language.Marlowe.Protocol.Load.Types (Node(..))
+import Language.Marlowe.Runtime.CLI.Monad (CLI)
+import Language.Marlowe.Runtime.ChainSync.Api (DatumHash)
+import Language.Marlowe.Runtime.Client (runMarloweLoadClient)
+import Network.TypedProtocol (N(..), Nat(..))
+import Options.Applicative (ParserInfo, help, info, metavar, progDesc, strArgument)
+import Plutus.V2.Ledger.Api (POSIXTime(..))
+import System.Exit (exitFailure)
+import System.FilePath (takeDirectory)
+import System.IO (hFlush, hPutStrLn, stderr)
+import Text.Printf (hPrintf)
+import UnliftIO (MonadUnliftIO, atomicModifyIORef, liftIO, newIORef)
+import UnliftIO.Directory (doesFileExist, makeAbsolute, withCurrentDirectory)
+
+newtype LoadCommand = LoadCommand
+  { contractFile :: FilePath
+  }
+
+loadCommandParser :: ParserInfo LoadCommand
+loadCommandParser = info parser $ progDesc "Load a contract into the runtime"
+  where
+    parser = LoadCommand
+      <$> contractFileOption
+    contractFileOption = strArgument $ mconcat
+      [ metavar "FILE_PATH"
+      , help "A file that contains the JSON representation of the contract to load."
+      ]
+
+runLoadCommand :: LoadCommand -> CLI ()
+runLoadCommand LoadCommand{..} = do
+  result <- runExceptT $ flip runReaderT [] do
+    nodeCount <- countNodes 0 contractFile
+    progress <- newIORef (-1 :: Int)
+    let
+      countWidth = floor (logBase 10 $ realToFrac nodeCount :: Double) + 1 :: Int
+      printStr = " [%-32s] %" <> show countWidth <> "d of " <> show nodeCount <> " nodes transferred.\r"
+      incrementProgress = do
+        newProgress <- atomicModifyIORef progress \i -> (i + 1, i + 1)
+        let
+          bar
+            | newProgress == 0 = ""
+            | newProgress == nodeCount = replicate 32 '='
+            | otherwise = reverse $ '>' : replicate ((newProgress * 32 `div` nodeCount) - 1) '='
+        liftIO $ hPrintf stderr printStr bar newProgress
+        liftIO $ hFlush stderr
+    lift $ ExceptT $ runMarloweLoadClient $ loadClient incrementProgress contractFile
+  liftIO case result of
+    Left err -> do
+      case err of
+        FileNotFound [] path -> do
+          hPrintf stderr "Cannot find contract file %s" path
+        FileNotFound (referencingFile : _) path -> do
+          hPrintf stderr "Cannot find contract file %s imported by %s" path referencingFile
+        FileInvalid referencingFile decodeError -> do
+          hPrintf stderr "Error in file %s:\n%s" referencingFile $ prettyPrintParseException decodeError
+        CyclicImport path -> do
+          hPrintf stderr "Cyclic import detected of file %s" path
+      hPutStrLn stderr ""
+      exitFailure
+    Right hash -> do
+      hPutStrLn stderr "\nDone."
+      putStrLn $ read $ show hash
+
+type CountM = ReaderT [FilePath] (ExceptT LoadError CLI)
+
+countNodes :: Int -> FilePath -> CountM Int
+countNodes count path = withContract path $ countNodes' count
+
+countNodes' :: Int -> LoadContract -> CountM Int
```

Comment:
I wonder if the multiplate in `Language.Marlowe.Core.V1.Plate` could be used to do this counting?

Reply:
Possibly, the concept of a `Node` is somewhat abstract though, and isn't directly captured by anything in there.

Reply:
Actually, I see that `LoadContract` is not `Contract`, which is what the plate acts on. :frowning_face: